subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/serialization/codecs/Codecs.kt

```diff
@@ -152,6 +153,7 @@ class Codecs(
         bind(ownerServiceCodec<BuildOperationListenerManager>())
         bind(ownerServiceCodec<BuildRequestMetaData>())
         bind(ownerServiceCodec<ListenerManager>())
+        bind(ownerServiceCodec<ClassLoaderCache>())
```

Comment:
If we're trying to serialise the class load cache, then the instant execution serialisation has gone much too far trying to serialise one of the other Gradle services. There should never be a need to serialise something that references the class loader cache, and we should instead serialise the outermost Gradle service that referenced by the task. Doing this means we avoid serializing a bunch of internal Gradle services that are already available when deserialising.

It would be good to figure out which service is actually being serialised here and handle that service instead.

Reply:
In this case it must have been for the test retry plugin using internals in
https://github.com/gradle/test-retry-gradle-plugin/blob/336dedb1dfd3b01e70709ff0816268e7fe378d70/plugin/src/main/java/org/gradle/testretry/internal/TestTaskConfigurer.java#L42

To get rid of a codec for `ClassLoaderCache` the test retry plugin must be changed.

Reply:
ugh. Then adding the codec is a reasonable change to make for now, I guess. We can later migrate the retry plugin to some public API. Let's go ahead and merge this PR.

Reply:
I think this could relatively easily be changed to pull this service lazily via that task's service registry. Would that resolve it?